bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/ResumingStreamingResultScanner.java

```diff
@@ -38,176 +32,74 @@
    */
-  public ResumingStreamingResultScanner(RetryOptions retryOptions, ReadRowsRequest originalRequest,
-      BigtableResultScannerFactory<FlatRow> scannerFactory, RpcMetrics rpcMetrics) {
-    this(retryOptions, originalRequest, scannerFactory, rpcMetrics, LOG);
+  public ResumingStreamingResultScanner(ResponseQueueReader responseQueueReader,
+      ScanHandler scanHandler) {
+    this.responseQueueReader = responseQueueReader;
+    this.scanHandler = scanHandler;
   }
 
-  @VisibleForTesting
-  ResumingStreamingResultScanner(
-      RetryOptions retryOptions,
-      ReadRowsRequest originalRequest,
-      BigtableResultScannerFactory<FlatRow> scannerFactory,
-      RpcMetrics rpcMetrics,
-      Logger logger) {
-    this.operationContext = rpcMetrics.timeOperation();
-    this.retryOptions = retryOptions;
-    this.requestManager =  new ReadRowsRequestManager(originalRequest);
-    this.scannerFactory = scannerFactory;
-    this.logger = logger;
-    this.rpcMetrics = rpcMetrics;
-
-    this.currentDelegate = scannerFactory.createScanner(originalRequest);
-    this.rpcContext = rpcMetrics.timeRpc();
+  /** {@inheritDoc} */
+  @Override
+  public final FlatRow[] next(int count) throws IOException {
+    ArrayList<FlatRow> resultList = new ArrayList<>(count);
+    for (int i = 0; i < count; i++) {
+      FlatRow row = next();
+      if (row == null) {
+        break;
+      }
+      resultList.add(row);
+    }
+    return resultList.toArray(new FlatRow[resultList.size()]);
   }
 
   /** {@inheritDoc} */
   @Override
   public FlatRow next() throws IOException {
     while (true) {
       try {
-        FlatRow result = currentDelegate.next();
+        Timer.Context timerContext = resultsTimer.time();
+        FlatRow result = responseQueueReader.getNextMergedRow();
         if (result != null) {
-          requestManager.updateLastFoundKey(result.getRowKey());
-          // We've had at least one successful RPC, reset the backoff and retry counter
-          currentErrorBackoff = null;
-          timeoutRetryCount = 0;
+          resultsMeter.mark();
         }
+        timerContext.close();
         return result;
       } catch (ScanTimeoutException rte) {
-        closeRpcContext();
-        closeCurrentDelegate();
-        ReadRowsRequest newRequest = handleScanTimeout(rte);
-        currentDelegate = scannerFactory.createScanner(newRequest);
-        this.rpcContext = rpcMetrics.timeRpc();
-      } catch (IOExceptionWithStatus ioe) {
-        closeRpcContext();
-        closeCurrentDelegate();
-        ReadRowsRequest newRequest = handleIOException(ioe);
-        currentDelegate = scannerFactory.createScanner(newRequest);
-        this.rpcContext = rpcMetrics.timeRpc();
+        scanHandler.handleTimeout(rte);
+      } catch (Throwable e) {
+        scanHandler.cancel();
+        throw new BigtableRetriesExhaustedException("Exhausted streaming retries.", e);
```

Comment:
Might be good to add a separate catch clause for `BigtableRetriesExhaustedException` to avoid double wrapping it